### ISSUE TITLE
core dump in regex with code and die/eval

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#7191]
Migrated from [rt.perl.org#27838](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=27838) (status was 'resolved')

Searchable as RT27838$

**Comment by p5pRT** [comment]
### From perl-5.8.0@ton.iguana.be

#### Created by perl-5.8.0@ton.iguana.be



This dumps core for me&#8203;:

\#\! /usr/bin/perl
@&#8203;tests = \(\{
&nbsp;              in => "ab\(c\(hi\)d\)ef\\2gh"\,
&nbsp;          \}\,
&nbsp;          \{
&nbsp;              in => "ab\(c\(hi\)d\)ef\\2gh"\,
&nbsp;          \}\);

for my $test \(@&#8203;tests\) \{
&nbsp;    eval \{ 
&nbsp;        $test\->\{in\} =~ /\(?\{die\}\)/;
&nbsp;    \};
\}

I only have 5\.8\.2\. I also let someone on 5\.8\.3 test it\, it didn't dump 
core for him\. However\, the core is pretty sensitive to the string lengths\.
By adding x 10 after each string his version cored too\, but for me that
makes the core go away\.
5\.8\.0 for i386\-bsd gives a memory fault\. Maybe a wild pointer\.\.\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.2:

Configured by ton at Sun Jan  4 19:19:06 CET 2004.

Summary of my perl5 (revision 5.0 version 8 subversion 2) configuration:
  Platform:
    osname=linux, osvers=2.6.0, archname=i686-linux-64int-ld
    uname='linux quasar 2.6.0 #3 thu dec 18 18:22:48 cet 2003 i686 gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=define
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -fomit-frame-pointer',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.4.0 20031231 (experimental)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.2:
    /usr/lib/perl5/5.8.2/i686-linux-64int-ld
    /usr/lib/perl5/5.8.2
    /usr/lib/perl5/site_perl/5.8.2/i686-linux-64int-ld
    /usr/lib/perl5/site_perl/5.8.2
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.2:
    HOME=/home/ton
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ton/bin.Linux:/home/ton/bin:/home/ton/bin.SampleSetup:/usr/local/bin:/usr/local/sbin:/usr/local/jre/bin:/home/oracle/product/9.0.1/bin:/usr/local/ar/bin:/usr/games/bin:/usr/X11R6/bin:/usr/share/bin:/usr/bin:/usr/sbin:/bin:/sbin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Mar 22\, 2004 at 03&#8203;:23&#8203;:25PM \-0000\, perl\-5\.8\.0@&#8203;ton\.iguana\.be \(via RT\) wrote&#8203;:
> \# New Ticket Created by  perl\-5\.8\.0@&#8203;ton\.iguana\.be 
> \# Please include the string&#8203;:  \[perl \#27838\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org&#8203;:80/rt3/Ticket/Display\.html?id=27838 >
> 
> 
> 
> This is a bug report for perl from perl\-5\.8\.0@&#8203;ton\.iguana\.be\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> This dumps core for me&#8203;:
> 
> \#\! /usr/bin/perl
> @&#8203;tests = \(\{
>               in => "ab\(c\(hi\)d\)ef\\2gh"\,
>           \}\,
>           \{
>               in => "ab\(c\(hi\)d\)ef\\2gh"\,
>           \}\);
> 
> for my $test \(@&#8203;tests\) \{
>     eval \{ 
>         $test\->\{in\} =~ /\(?\{die\}\)/;
>     \};
> \}
> 
> I only have 5\.8\.2\. I also let someone on 5\.8\.3 test it\, it didn't dump 
> core for him\. However\, the core is pretty sensitive to the string lengths\.
> By adding x 10 after each string his version cored too\, but for me that
> makes the core go away\.
> 5\.8\.0 for i386\-bsd gives a memory fault\. Maybe a wild pointer\.\.\.
>     SHELL=/bin/bash

Fails in bleed too\. The die/eval doesn't restore PL\_comppad/PL\_curpad to
point to main's pad rather then the regex\-code's pad\. Tragedy soon follows\.

Looks like something to add to my "Big List of things to fix with
/\(?\{\.\.\.\}\)/"\.


Dave\.


\-\- 
"Foul and greedy Dwarf \- you have eaten the last candle\."
&nbsp;    \-\- "Hordes of the Things"\, BBC Radio\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From buu@erxz.com

The following code generates a segfault&#8203;:
eval \{ /\(?\{die\}\)/ \}; $@&#8203;

Note that the $@&#8203; is necessary as it doesn't segfault without it\.
If you omit the reference to $@&#8203; it doesn't segfault but it doesn't
appear to execute any code after the eval block\. For example\,
perl \-e"eval \{ /\(?\{die\}\)/ \}; print 'foo' " 
produces no output\, but doesn't segfault\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.9.1:

Configured by Administrator at Thu Dec 30 13:35:08 2004.

Summary of my perl5 (revision 5 version 9 subversion 1) configuration:
  Platform:
    osname=MSWin32, osvers=4.0, archname=MSWin32-x86-multi-
thread
    uname=''
    config_args='undef'
    hint=recommended, useposix=true, d_sigaction=undef
    usethreads=define useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cl', ccflags ='-nologo -Gf -W3 -MD -DNDEBUG -O1 -DWIN32 -
D_CONSOLE -DNO_STRICT -DHAVE_DES_FCRYPT  -
DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -
DUSE_PERLIO -DPERL_MSVCRT_READFIX',
    optimize='-MD -DNDEBUG -O1',
    cppflags='-DWIN32'
    ccversion='', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=undef, longlongsize=8, d_longdbl=define, longdblsize=10
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='__int64', 
lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='link', ldflags ='-nologo -nodefaultlib -release  -
libpath:"c:\perl\lib\CORE"  -machine:x86'
    libpth=\lib
    libs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib 
uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib 
odbccp32.lib msvcrt.lib
    perllibs=  oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  
comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib 
uuid.lib wsock32.lib mpr.lib winmm.lib  version.lib odbc32.lib 
odbccp32.lib msvcrt.lib
    libc=msvcrt.lib, so=dll, useshrplib=yes, libperl=perl59.lib
    gnulibc_version='undef'
  Dynamic Linking:
    dlsrc=dl_win32.xs, dlext=dll, d_dlsymun=undef, ccdlflags=' '
    cccdlflags=' ', lddlflags='-dll -nologo -nodefaultlib -release  -
libpath:"c:\perl\lib\CORE"  -machine:x86'

Locally applied patches:
    


@INC for perl v5.9.1:
    C:/perl/lib
    C:/perl/site/lib
    .


Environment for perl v5.9.1:
    HOME (unset)
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    
PATH=C:\WINNT\system32;C:\WINNT;C:\WINNT\System32\Wbem;C:\
perl\bin;C:\Program Files\Microsoft Visual 
Studio\Common\Tools\WinNT;C:\Program Files\Microsoft Visual 
Studio\Common\MSDev98\Bin;C:\Program Files\Microsoft Visual 
Studio\Common\Tools;C:\Program Files\Microsoft Visual 
Studio\VC98\bin
    PERL_BADLANG (unset)
    SHELL (unset)


```
</details>

**Comment by p5pRT** [comment]
### From porton@ex-code.com

#### Created by porton@ex-code.com



$ perl \-e 'eval \{ /\(?\{die\}\)/ \}; print'
Segmentation fault

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

Site configuration information for perl v5.8.4:

Configured by Debian Project at Mon Oct 25 01:52:37 EST 2004.

Summary of my perl5 (revision 5 version 8 subversion 4) configuration:
  Platform:
    osname=linux, osvers=2.4.27-ti1211, archname=i386-linux-thread-multi
    uname='linux kosh 2.4.27-ti1211 #1 sun sep 19 18:17:45 est 2004 i686 gnulinux '
    config_args='-Dusethreads -Duselargefiles -Dccflags=-DDEBIAN -Dcccdlflags=-fPIC -Darchname=i386-linux -Dprefix=/usr -Dprivlib=/usr/share/perl/5.8 -Darchlib=/usr/lib/perl/5.8 -Dvendorprefix=/usr -Dvendorlib=/usr/share/perl5 -Dvendorarch=/usr/lib/perl5 -Dsiteprefix=/usr/local -Dsitelib=/usr/local/share/perl/5.8.4 -Dsitearch=/usr/local/lib/perl/5.8.4 -Dman1dir=/usr/share/man/man1 -Dman3dir=/usr/share/man/man3 -Dsiteman1dir=/usr/local/man/man1 -Dsiteman3dir=/usr/local/man/man3 -Dman1ext=1 -Dman3ext=3perl -Dpager=/usr/bin/sensible-pager -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dlibperl=libperl.so.5.8.4 -Dd_dosuid -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -DTHREADS_HAVE_PIDS -DDEBIAN -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='3.3.5 (Debian 1:3.3.5-1)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -lgdbm_compat -ldb -ldl -lm -lpthread -lc -lcrypt
    perllibs=-ldl -lm -lpthread -lc -lcrypt
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=true, libperl=libperl.so.5.8.4
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.4:
    /usr/lib/perl5/5.005/i386-linux
    /etc/perl
    /usr/local/lib/perl/5.8.4
    /usr/local/share/perl/5.8.4
    /usr/lib/perl5
    /usr/share/perl5
    /usr/lib/perl/5.8
    /usr/share/perl/5.8
    /usr/local/lib/site_perl
    .


Environment for perl v5.8.4:
    HOME=/home/porton
    LANG (unset)
    LANGUAGE=en_US
    LC_COLLATE=ru_RU.KOI8-R
    LC_CTYPE=ru_RU.KOI8-R
    LC_MESSAGES=en_US
    LC_MONETARY=ru_RU.KOI8-R
    LC_NUMERIC=C
    LC_TIME=C
    LD_LIBRARY_PATH=/usr/local/lib
    LOGDIR (unset)
    PATH=~/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/sbin:/usr/sbin:/usr/local/sbin
    PERL5LIB=/usr/lib/perl5/5.005/i386-linux
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @tux

On 25 Mar 2005 06&#8203;:04&#8203;:45 \-0000\, "Victor Porton\,\,\," \(via RT\)
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> $ perl \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> Segmentation fault

This has been solved between 5\.8\.5 and 5\.8\.6&#8203;:

pc09&#8203;:/home/merijn 103 > perl5\.8\.5 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
Segmentation fault
Exit 139
pc09&#8203;:/home/merijn 104 > perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
pc09&#8203;:/home/merijn 105 >

So if this is in anyway a showstopper for you\, I'd advise you to
upgrade your 5\.8\.4 to 5\.8\.6


\-\- 
H\.Merijn Brand        Amsterdam Perl Mongers \(http&#8203;://amsterdam\.pm\.org/\)
using Perl 5\.6\.2\, 5\.8\.0\, 5\.8\.5\, & 5\.9\.2  on HP\-UX 10\.20\, 11\.00 & 11\.11\,
 AIX 4\.3 & 5\.2\, SuSE 9\.1 & 9\.2\, and Cygwin\. http&#8203;://www\.cmve\.net/~merijn
Smoking perl&#8203;: http&#8203;://www\.test\-smoke\.org\,    perl QA&#8203;: http&#8203;://qa\.perl\.org
 reports  to&#8203;: smokers\-reports@&#8203;perl\.org\,                perl\-qa@&#8203;perl\.org

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From at@altlinux.ru

On Fri\, Mar 25\, 2005 at 09&#8203;:31&#8203;:56AM \+0100\, H\.Merijn Brand wrote&#8203;:
> > $ perl \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > Segmentation fault
> 
> This has been solved between 5\.8\.5 and 5\.8\.6&#8203;:
> 
> pc09&#8203;:/home/merijn 103 > perl5\.8\.5 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> Segmentation fault
> Exit 139
> pc09&#8203;:/home/merijn 104 > perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> pc09&#8203;:/home/merijn 105 >
> 
> So if this is in anyway a showstopper for you\, I'd advise you to
> upgrade your 5\.8\.4 to 5\.8\.6

I can reproduce this with 5\.8\.6 and 5\.9\.2 \(as of change \#24046\)&#8203;:

$ perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
zsh&#8203;: segmentation fault  perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
$ perl5\.9\.2 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
zsh&#8203;: segmentation fault  perl5\.9\.2 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
$

**Comment by p5pRT** [comment]
### From @demerphq

On Mar 31\, 2005 11&#8203;:27 PM\, Alexey Tourbin \<at@&#8203;altlinux\.ru> wrote&#8203;:
> On Fri\, Mar 25\, 2005 at 09&#8203;:31&#8203;:56AM \+0100\, H\.Merijn Brand wrote&#8203;:
> > > $ perl \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > > Segmentation fault
> >
> > This has been solved between 5\.8\.5 and 5\.8\.6&#8203;:
> >
> > pc09&#8203;:/home/merijn 103 > perl5\.8\.5 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > Segmentation fault
> > Exit 139
> > pc09&#8203;:/home/merijn 104 > perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > pc09&#8203;:/home/merijn 105 >
> >
> > So if this is in anyway a showstopper for you\, I'd advise you to
> > upgrade your 5\.8\.4 to 5\.8\.6
> 
> I can reproduce this with 5\.8\.6 and 5\.9\.2 \(as of change \#24046\)&#8203;:
> 
> $ perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> zsh&#8203;: segmentation fault  perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> $ perl5\.9\.2 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> zsh&#8203;: segmentation fault  perl5\.9\.2 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> $

Are you saying that blead prior to 24046 doesnt have this problem? 

Cheers\,
Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @demerphq

On Mar 31\, 2005 11&#8203;:27 PM\, Alexey Tourbin \<at@&#8203;altlinux\.ru> wrote&#8203;:
> On Fri\, Mar 25\, 2005 at 09&#8203;:31&#8203;:56AM \+0100\, H\.Merijn Brand wrote&#8203;:
> > > $ perl \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > > Segmentation fault
> >
> > This has been solved between 5\.8\.5 and 5\.8\.6&#8203;:
> >
> > pc09&#8203;:/home/merijn 103 > perl5\.8\.5 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > Segmentation fault
> > Exit 139
> > pc09&#8203;:/home/merijn 104 > perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > pc09&#8203;:/home/merijn 105 >
> >
> > So if this is in anyway a showstopper for you\, I'd advise you to
> > upgrade your 5\.8\.4 to 5\.8\.6
> 
> I can reproduce this with 5\.8\.6 and 5\.9\.2 \(as of change \#24046\)&#8203;:
> 
> $ perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> zsh&#8203;: segmentation fault  perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> $ perl5\.9\.2 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> zsh&#8203;: segmentation fault  perl5\.9\.2 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> $

If you omit the print it doesn't segfault with blead as of  20487 and
the segfault is in pp\_gvsv\(\) at the line where it says

&nbsp;  PUSHs\(GvSV\(cGVOP\_gv\)\);

so it looks like the die breaks something outside of the regex\.

yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @iabyn

On Fri\, Apr 01\, 2005 at 10&#8203;:02&#8203;:59AM \+0200\, demerphq wrote&#8203;:
> > I can reproduce this with 5\.8\.6 and 5\.9\.2 \(as of change \#24046\)&#8203;:
> > 
> > $ perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > zsh&#8203;: segmentation fault  perl5\.8\.6 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > $ perl5\.9\.2 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > zsh&#8203;: segmentation fault  perl5\.9\.2 \-e 'eval \{ /\(?\{die\}\)/ \}; print'
> > $
> 
> If you omit the print it doesn't segfault with blead as of  20487 and
> the segfault is in pp\_gvsv\(\) at the line where it says
> 
>   PUSHs\(GvSV\(cGVOP\_gv\)\);
> 
> so it looks like the die breaks something outside of the regex\.

die within \(?\{\.\.\.\}\) is just plain broken at the moment due to the way the
block is invoked\. Whether it coredumps with specific releases is
coincidental\. Its on my list of things to fix\.

\-\- 
"But Sidley Park is already a picture\, and a most amiable picture too\.
The slopes are green and gentle\. The trees are companionably grouped at
intervals that show them to advantage\. The rill is a serpentine ribbon
unwound from the lake peaceably contained by meadows on which the right
amount of sheep are tastefully arranged\." \-\- Lady Croom \- Arcadia

**Comment by p5pRT** [comment]
### From @smpeters

> \[buu@&#8203;erxz\.com \- Tue Jan 11 18&#8203;:15&#8203;:56 2005\]&#8203;:
> 
> The following code generates a segfault&#8203;:
> eval \{ /\(?\{die\}\)/ \}; $@&#8203;
> 
> Note that the $@&#8203; is necessary as it doesn't segfault without it\.
> If you omit the reference to $@&#8203; it doesn't segfault but it doesn't
> appear to execute any code after the eval block\. For example\,
> perl \-e"eval \{ /\(?\{die\}\)/ \}; print 'foo' " 
> produces no output\, but doesn't segfault\.
> 
> 

This is repeatable with the current blead\.  Backtrace and \-DstvP output
below\.

\#0  0x080e6902 in Perl\_pp\_gvsv \(my\_perl=0x81b6008\) at pp\_hot\.c&#8203;:61
61              PUSHs\(GvSV\(cGVOP\_gv\)\);
\(gdb\) bt
\#0  0x080e6902 in Perl\_pp\_gvsv \(my\_perl=0x81b6008\) at pp\_hot\.c&#8203;:61
\#1  0x080cbe61 in Perl\_runops\_debug \(my\_perl=0x81b6008\) at dump\.c&#8203;:1600
\#2  0x08066c3f in S\_run\_body \(my\_perl=0x81b6008\, oldscope=1\) at perl\.c&#8203;:2038
\#3  0x080666b5 in perl\_run \(my\_perl=0x81b6008\) at perl\.c&#8203;:1972
\#4  0x0806046d in main \(argc=2\, argv=0xbffffab4\, env=0xbffffac0\) at
perlmain\.c&#8203;:103

steve@&#8203;kirk&#8203;:\~/perl\-current$ \./perl \-DstvP \-wle'eval \{ /\(?\{die\}\)/ \}; print $@&#8203;'
Use of uninitialized value\.
filter\_add func 806ce46 \(\)
Use of uninitialized value\.
filter\_read 0&#8203;: via function 81b6d88 \(\)
Use of uninitialized value at \-e line 1\.
filter\_read 0&#8203;: via function 81b6d88 \(\)
filter\_del func 806ce46
EXECUTING\.\.\.


STACK 0&#8203;: MAIN

\(\-e&#8203;:0\)  enter

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 

\(\-e&#8203;:0\)  nextstate

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 

\(\-e&#8203;:1\)  entertry

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 
&nbsp;  CX 1&#8203;: EVAL   => 
&nbsp;  retop=nextstate

\(\-e&#8203;:1\)  nextstate

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 
&nbsp;  CX 1&#8203;: EVAL   => 
&nbsp;  retop=nextstate

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 
&nbsp;  CX 1&#8203;: EVAL   => 
&nbsp;  retop=nextstate

\(\-e&#8203;:1\)  match
Use of uninitialized value $\_ in pattern match \(m//\) at \-e line 1\.

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 
&nbsp;  CX 1&#8203;: EVAL   => 
&nbsp;  retop=nextstate

\(\-e&#8203;:1\)  nextstate

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 
&nbsp;  CX 1&#8203;: EVAL   => 
&nbsp;  retop=nextstate

\(\(re\_eval 1\)&#8203;:1\) pushmark

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 
&nbsp;  CX 1&#8203;: EVAL   => \*  
&nbsp;  retop=nextstate

\(\(re\_eval 1\)&#8203;:1\) die

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 

\(\-e&#8203;:1\)  nextstate

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => 

\(\-e&#8203;:1\)  pushmark

STACK 0&#8203;: MAIN
&nbsp;  CX 0&#8203;: BLOCK  => \*  

\(\-e&#8203;:1\)  gvsvSegmentation fault \(core dumped\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @smpeters

> \[perl\-5\.8\.0@&#8203;ton\.iguana\.be \- Mon Mar 22 07&#8203;:23&#8203;:24 2004\]&#8203;:
> 
> 
> This is a bug report for perl from perl\-5\.8\.0@&#8203;ton\.iguana\.be\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.2\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> This dumps core for me&#8203;:
> 
> \#\! /usr/bin/perl
> @&#8203;tests = \(\{
>               in => "ab\(c\(hi\)d\)ef\\2gh"\,
>           \}\,
>           \{
>               in => "ab\(c\(hi\)d\)ef\\2gh"\,
>           \}\);
> 
> for my $test \(@&#8203;tests\) \{
>     eval \{
>         $test\->\{in\} =~ /\(?\{die\}\)/;
>     \};
> \}
> 
> I only have 5\.8\.2\. I also let someone on 5\.8\.3 test it\, it didn't dump
> core for him\. However\, the core is pretty sensitive to the string
>    lengths\.
> By adding x 10 after each string his version cored too\, but for me
>    that
> makes the core go away\.
> 5\.8\.0 for i386\-bsd gives a memory fault\. Maybe a wild pointer\.\.\.
> 

The above code is currently coredumping with the following backtrace\. 
\#0  0x080d6da3 in Perl\_vivify\_ref \(sv=0x19\, to\_what=64\) at pp\_hot\.c&#8203;:2989
2989        SvGETMAGIC\(sv\);
\(gdb\) bt
\#0  0x080d6da3 in Perl\_vivify\_ref \(sv=0x19\, to\_what=64\) at pp\_hot\.c&#8203;:2989
\#1  0x080c44fa in Perl\_pp\_padsv \(\) at pp\_hot\.c&#8203;:279
\#2  0x0808b967 in Perl\_runops\_debug \(\) at dump\.c&#8203;:1698
\#3  0x080b8c15 in S\_run\_body \(oldscope=1\) at perl\.c&#8203;:2373
\#4  0x080b823f in perl\_run \(my\_perl=0x826e008\) at perl\.c&#8203;:2298
\#5  0x0805e700 in main \(argc=2\, argv=0xbf97f534\, env=0xbf97f540\)
&nbsp;    at perlmain\.c&#8203;:103

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Apr 17\, 2006 at 11&#8203;:08&#8203;:54AM \-0700\, Steve Peters via RT wrote&#8203;:
> The above code is currently coredumping with the following backtrace\. 

Yeah\, re\_evals still can't handle exceptions safely\.
It's on Dave's Big List of Things to Do\.

\-\- 
"You're so sadly neglected\, and often ignored\.
A poor second to Belgium\, When going abroad\."
&nbsp;    \-\- Monty Python \- "Finland"

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
